Support expressions in filters

anandthakker opened the pull request:
Closes #4077
Closes #4410

- Converts classic filters to expressions under the hood
 - Also accepts `{expression: [...]}`, requiring that the given
 expression be of type Boolean
 - [ ] Benchmark to assess performance impact acainst classic filters
 - [ ] Add fixtures for `validation.test.js`
 - [ ] Manually test debug page
 - [ ] Add some kind of partial string matching? ( https://github.com/mapbox/mapbox-gl-js/issues/4089 / https://github.com/mapbox/mapbox-gl-js/issues/4113 )

 cc @jfirebaugh @aparlato

indus commented:
Will there be an expression for "hasOwnProperty". I'm living with a patched version since #4490

anandthakker commented:
@indus Good question - there is a `["has", "propertyName"]` expression (see https://github.com/mapbox/mapbox-gl-js/pull/4777), but we should also add `[ "contains" , ... ]` for checking presence in an array.

anandthakker commented:
@aparlato @nickidlugash updated with:
 - support for `[zoom]`
 - `[contains]`

Still working on performance but it should be usable for developing styles.

anandthakker commented:
Rebased onto master and reworked to take advantage of #5234 -- performance is much better, but evaluation performance is still about 8x compared to non-expressions-based filters:

<img width="760" alt="screen shot 2017-09-28 at 3 52 25 pm" src="https://user-images.githubusercontent.com/4974151/30987452-0b6e53ae-a465-11e7-8eeb-a9c1b17f6d91.png">

And creation performance is much slower:
<img width="766" alt="screen shot 2017-09-28 at 3 52 10 pm" src="https://user-images.githubusercontent.com/4974151/30987464-1661488e-a465-11e7-8d21-e3c4215a3aa0.png">

anandthakker commented:
Along the lines of #5379 , @mourner suggests leaving the existing filter implementation in place, and making expressions support additive / opt-in.

anandthakker commented:
Back to being a bit worse after rebasing onto the recent updates on master (no-eval, etc.):

Bench | master | expressions-filter | gist
------ | -- | -- | --
FilterCreate | 0.303 ms | 9.81 ms ( +9.51 ms / 15.1 std devs ) | https://bl.ocks.org/anonymous/raw/f98f920d15e3df367eb5175987548939/ 
FilterEvaluate | 0.136 ms | 2.93 ms ( +2.79 ms / 37.3 std devs ) | https://bl.ocks.org/anonymous/raw/b42d635f2eda75b0627f95801985d7ba/

anandthakker commented:
cc @mourner @jfirebaugh I was hoping the recent changes would mean we could stick with converting old-style filters, but looks like we should drop the conversion for now.

jfirebaugh commented:
https://github.com/mapbox/mapbox-gl-js/compare/expressions-filter-perf gets it down to 3.5-4x slower.

jfirebaugh commented:
I merged the alternative without conversion, #5434. We'll hopefully be resurrecting the conversion code from this PR at some point, but closing for now.
